File: api/app/Models/PronunciationScore.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class PronunciationScore extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id',
        'tongue_twister_id',
        'overall_score',
        'phoneme_scores',
        'feedback',
        'recording_path',
        'analyzed_at',
    ];

    protected $casts = [
        'overall_score' => 'decimal:2',
        'phoneme_scores' => 'array',
        'feedback' => 'array',
        'analyzed_at' => 'datetime',
    ];

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function tongueTwister(): BelongsTo
    {
        return $this->belongsTo(TongueTwister::class);
    }

    public function getScoreGradeAttribute(): string
    {
        return match(true) {
            $this->overall_score >= 90 => 'Excellent',
            $this->overall_score >= 80 => 'Good',
            $this->overall_score >= 70 => 'Fair',
            $this->overall_score >= 60 => 'Needs Work',
            default => 'Keep Practicing',
        };
    }

    public function getScoreColorAttribute(): string
    {
        return match(true) {
            $this->overall_score >= 90 => '#4CAF50',
            $this->overall_score >= 80 => '#8BC34A',
            $this->overall_score >= 70 => '#FFC107',
            $this->overall_score >= 60 => '#FF9800',
            default => '#F44336',
        };
    }
}

File: apk/lib/features/sl/models/pronunciation_score.dart
import 'package:equatable/equatable.dart';

class PronunciationScore extends Equatable {
  final double overallScore;
  final Map<String, double> phonemeScores;
  final List<String> feedback;
  final String recordingPath;
  final DateTime timestamp;

  const PronunciationScore({
    required this.overallScore,
    required this.phonemeScores,
    required this.feedback,
    required this.recordingPath,
    required this.timestamp,
  });

  factory PronunciationScore.fromJson(Map<String, dynamic> json) {
    return PronunciationScore(
      overallScore: (json['overall_score'] as num).toDouble(),
      phonemeScores: Map<String, double>.from(
        json['phoneme_scores'].map((key, value) => MapEntry(key, (value as num).toDouble()))
      ),
      feedback: List<String>.from(json['feedback']),
      recordingPath: json['recording_path'] ?? '',
      timestamp: DateTime.parse(json['timestamp']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'overall_score': overallScore,
      'phoneme_scores': phonemeScores,
      'feedback': feedback,
      'recording_path': recordingPath,
      'timestamp': timestamp.toIso8601String(),
    };
  }

  String get scoreGrade {
    if (overallScore >= 90) return 'Excellent';
    if (overallScore >= 80) return 'Good';
    if (overallScore >= 70) return 'Fair';
    if (overallScore >= 60) return 'Needs Work';
    return 'Keep Practicing';
  }

  String get scoreColor {
    if (overallScore >= 90) => '#4CAF50';
    if (overallScore >= 80) => '#8BC34A';
    if (overallScore >= 70) => '#FFC107';
    if (overallScore >= 60) => '#FF9800';
    return '#F44336';
  }

  @override
  List<Object?> get props => [overallScore, phonemeScores, feedback, recordingPath, timestamp];
}
